\documentclass[
10pt, % Main document font size
a4paper, % Paper type, use 'letterpaper' for US Letter paper
oneside, % One page layout (no page indentation)
%twoside, % Two page layout (page indentation for binding and different headers)
headinclude,footinclude, % Extra spacing for the header and footer
BCOR5mm, % Binding correction
]{scrartcl}

\input{structure.tex}
\usepackage{adjustbox}
\hyphenation{Fortran hy-phen-ation} 
\captionsetup{font=footnotesize}

%----------------------------------------------------------------------------------------
%	TITLE AND AUTHOR(S)
%----------------------------------------------------------------------------------------
\title{\normalfont\spacedallcaps{Article Title}} % The article title

\author{\spacedlowsmallcaps{Fatemeh Hadi Nezhad\textsuperscript{1}}}

\date{2019} % An optional date to appear under the author(s)

%----------------------------------------------------------------------------------------

\begin{document}

%----------------------------------------------------------------------------------------
%	HEADERS
%----------------------------------------------------------------------------------------

\renewcommand{\sectionmark}[1]{\markright{\spacedlowsmallcaps{#1}}} % The header for all pages 
\lehead{\mbox{\llap{\small\thepage\kern1em\color{halfgray} \vline}\color{halfgray}\hspace{0.5em}\rightmark\hfil}} % The header style

\pagestyle{scrheadings} % Enable the headers specified in this block

%----------------------------------------------------------------------------------------
%	TABLE OF CONTENTS & LISTS OF FIGURES AND TABLES
%----------------------------------------------------------------------------------------

\maketitle % Print the title/author/date block

\setcounter{tocdepth}{2} % Set the depth of the table of contents to show sections and subsections only

\tableofcontents % Print the table of contents

%----------------------------------------------------------------------------------------
%	AUTHOR AFFILIATIONS
%----------------------------------------------------------------------------------------

\let\thefootnote\relax\footnotetext{* \textit{Department of Quantitative System Biology, University of California, Merced, United States}}

%----------------------------------------------------------------------------------------

\newpage 

%----------------------------------------------------------------------------------------
%	Preparing Data
%----------------------------------------------------------------------------------------

\section{Preparing Data}
%------------------------------------------------
\textbf{TryTryp genome data}. From tritrypdb website, we downloaded the version 41 of 46 TryTryp genomes released on 2018-12-05. Genomes are compared based on number of sequence fragments relative to their length as shown in figure ~\vref{fig:gallery}. 
All the results, scripts and input data for TryTryp version 41 can be found \href{thesentence}{here}. 

\begin{figure}[tb]
\centering 
\includegraphics[width=0.8\columnwidth]{GenomeComparisonV41.png} 
\caption[Genome Comparison]{Comparing genomes based on formula ${(\frac{f(x)}{\max(f(x))})}^{-1}$ which f(x) = Number of sequences in genome devided by the length of genome. the higher the bar is the better genome is sequenced.} % The text in the square bracket is the caption for the list of figures while the text in the curly brackets is the figure caption
\label{fig:gallery} 
\end{figure}

\subsection{\textbf{tRNA gene annotation}}
\subsubsection*{tRNA gene prediction}
In order to annotate the tRNA genes for the sequenced TryTryp genomes, we used two computational methods for tRNA prediction, tRNAscan-SE and Aragorn. We integrated the result of both genefinders by keeping the union of tRNA gene predictions generated by tRNAscan-SE v2.0 using defult options (Lowe and Eddy 1997) and Aragorn v1.2.38 using options -i116 -t -br -seq -w -e -l -d (Laslett and Canback 2004). Genes with overlapped coordinate were considered one gene. However, the identity and exact coordinate of both genefinders we saved seperately to be compared later. 

%----------------------------------------------------------------------------------------
%	Initiator tRNA prediction
%----------------------------------------------------------------------------------------

\subsubsection*{Initiator tRNA prediction}
We predicted the initiator tRNAs for the genes with anticodon 'CAT' from intersection of both tRNAscan (TSE) and Aragorn (ARA) Based on Conserved positions of initiators in Eukarya from the study by CHRISTIAN MARCK and HENRI GROSJEAN. Based on the fallowing citeria:
\begin{enumerate}[noitemsep] % [noitemsep] removes whitespace between the items for a compact look
\item In all eukaryotic tDNA-iMet, positions 11–24 are occupied by C-G, However, eukaryotic elongators
also prefer C-G at these positions.

\item Initiator tDNAs from Eukarya use A54 and A60. Some eukaryotic elongators also use either A54 or
A60 but none (with only one exception) uses both

\item Initiator tDNA-iMet (CAT) from all domains display the GGG sequence (Mandal et al+, 1996) or,
very seldom, the AGG sequence at positions 29 to 31, pairing with the complementary CCC or
CCT sequences at positions 39 to 41

\item Another domain-specific feature in all eukaryotic initiators is the systematic nonoccupancy of all
optional positions of the D-loop (17, 17a, 20a, and 20b) whereas in elongators, only position 17a is always unoccupied.

\item At position 20, A is strictly conserved in all eukaryotic initiators
\end{enumerate} 
To investigate all these features we clustered CAT tRNA genes using Levenshtein (edit) distance between gene sequences and Ward.D2 method to measure the dissimilarity between each two clusters. We ended up with three clusters. Table \ref{table:1} investigates each of these features in each column. from this table we see that only tRNA genes in cluster 1 have almost all the conserved features for eukaryotic initiators.


\begin{table}[hbt]
\caption{Table of CAT clusters to show how many tRNA genes in each cluster satisfy each feature}
\begin{adjustbox}{width=\columnwidth,center}
%\caption{Table of Grades}
%\centering
\begin{tabular}{|l|lllllllll|}
\hline
Clusters & \# tRNAs & 11–24(C-G) & 54-60(A-A)(T-T) & 1-72(A-T) & 29-31(GGG) & 39-41(CCC/CCT) & \#  posisInDloop & 20A & distanceRange \\
\hline
Cluster1 & 76 & 76 & 76 & 76 & 76 & 76 & 7 & 75 & 0-6\\
Cluster2 & 95 & 95 & 2 & 0 & 95 & 95 & 8 & 0 & 0-8\\
Cluster3 & 2 & 2 & 2 & 0 & 0 & 0 & 8/9 & 0 & 0-22\\
\hline
\end{tabular}
\label{table:1}
\end{adjustbox}
\end{table}



%----------------------------------------------------------------------------------------
%	tRNA Genes Integration
%----------------------------------------------------------------------------------------

\subsection{\textbf{Summary of predicted TryTryp tRNA genes}}

To investigate and compare tRNA genes predicted by two gene finders TSE and ARA, we made four sets of genes. Set one, TSE and ARA intersection, Set two, TSE and ARA union, Set three, genes found by ARA and Set four, genes found by TSE . for intersection set, we dismissed genes which had different identity by ARA and TSE. for union set, we picked TSE identity over ARA. Table  ~\ref{table:2} shows a summary of these four sets. Further, to compare the coordinates of genes annotated by ARA and TSE we made a heatmap shown in figure ~\ref{fig:heatmap}. We see from this figure that ... 
%----------------------------------------------------------------------------------------
%	???? Analyse why we have the differences ....
%----------------------------------------------------------------------------------------

Figure ~\ref{fig:counts} shows the number of genes annotated for each genome, and Figure ~\ref{fig:types} shows which tRNA functional classes were not annotated by both TSE and ARA for each genome. 

\begin{table}[hbt]
\caption{summary of the predicted genes by TSE and ARA. pseudo genes are marked as \$, initiators as X, stop as \#, sup as "?", sec as Z and pyl as O}
\begin{adjustbox}{width=\columnwidth,center}
%\caption{Table of Grades}
%\centering
% make it as two table
\begin{tabular}{|l|lllllllllllllllllllllllllllllllllll|}
\hline
GeneSet & \# tRNA & \# nucleotides & N/T & gene length & \%G & \%C & \%T & \%A & \%intron & A & C & D & E & F & G & H & I & K & L & M & N & P & Q & R & S & T & V & W & Y & X & Z & \$ & ? & \# & O\\
TSE2 & 3629 & 270211 & 74.46 & 50-164 & 31.99 & 26.11 & 23.22 & 18.68 &  2.618 & 214 & 64 & 104 & 162 & 110 & 234 &  80 & 179 & 190 & 338 & 108 & 126 & 201 & 162 & 350 & 238 & 219 & 241 & 52 &  94 & 76 & 78 & 28 & 3 & 0 & 0\\
ARA & 4345 & 372392 & 85.71 & 70-215 & 32.64 & 26.92 & 22.87 & 17.57 & 14.684 & 257 & 86 & 123 & 192 & 125 & 339 & 129 & 213 & 194 & 393 & 101 & 153 & 228 & 175 & 420 & 362 & 248 & 282 & 60 &  90 & 76 & 82 &  0 & 0 & 2 & 4\\
UNION & 4379 & 377587 & 86.23 & 50-215 & 32.81 & 26.66 & 22.87 & 17.65 & 15.346 & 259 & 86 & 118 & 193 & 130 & 344 & 129 & 220 & 197 & 380 & 112 & 143 & 229 & 175 & 421 & 369 & 249 & 282 & 57 & 106 & 76 & 82 & 28 & 3 & 2 & 2\\
INTERSECTION & 3560 & 265016 & 74.44 & 68-89 & 32.01 & 26.13 & 23.22 & 18.64 &  2.331 & 212 & 64 & 104 & 161 & 105 & 229 &  80 & 172 & 187 & 338 &  97 & 125 & 200 & 162 & 349 & 230 & 218 & 241 & 52 &  78 & 76 & 78 &  6 & 0 & 0 & 0\\
\hline
\end{tabular}
\label{table:2}
\end{adjustbox}
\end{table}


\begin{figure}[tb]
\centering 
\includegraphics[width=0.8\columnwidth]{EndDisplacement.png} 
\caption[Genome Comparison]{Empirical joint distribution of end-displacements in ? tRNA gene models found by both Aragorn and tRNAscan-SE 2.0.} % The text in the square bracket is the caption for the list of figures while the text in the curly brackets is the figure caption
\label{fig:heatmap} 
\end{figure}
 
\begin{figure}[tb]
\centering 
\includegraphics[width=0.8\columnwidth]{intersecttRNAcounts.png} 
\caption[Number of Genes annotated]{Number of genes annotated by both TSE and ARA for each TryTryp genome.} % The text in the square bracket is the caption for the list of figures while the text in the curly brackets is the figure caption
\label{fig:counts} 
\end{figure}

\begin{figure}[tb]
\centering 
\includegraphics[width=0.8\columnwidth]{ara_funcPerc.png} 
\caption[Genome Comparison]{Percentage of 22 tRNA types annotated by both TSE and ARA for each TryTryp genomes. The labels on top of each bar shows which tRNA classes are not annotated for the genome.} % The text in the square bracket is the caption for the list of figures while the text in the curly brackets is the figure caption
\label{fig:types} 
\end{figure}




....
%----------------------------------------------------------------------------------------
%	BIBLIOGRAPHY
%----------------------------------------------------------------------------------------
\newpage
\renewcommand{\refname}{\spacedlowsmallcaps{References}} % For modifying the bibliography heading

\bibliographystyle{unsrt}

\bibliography{sample.bib} % The file containing the bibliography

%----------------------------------------------------------------------------------------

\end{document}